service/history/shard/context_impl.go

```diff
@@ -684,6 +686,35 @@ func (s *ContextImpl) AppendHistoryEvents(
 	return size, err0
 }
 
+func (s *ContextImpl) DeleteWorkflowExecution(
+	curRequest *persistence.DeleteCurrentWorkflowExecutionRequest,
+	delRequest *persistence.DeleteWorkflowExecutionRequest,
+) error {
+	if err := s.errorByState(); err != nil {
+		return err
+	}
+
+	curRequest.ShardID = s.shardID
+	delRequest.ShardID = s.shardID
```

Comment:
caller should set the shard ID

Reply:
Hah.. I asked that here: https://github.com/temporalio/temporal/pull/2136#discussion_r743204197
Why not let the shard set it for all the methods? (I was going to make that change)

Reply:
we can, previously the shard ID is set within the execution persistence layer.
but when will the request contain all necessary information?

Reply:
As early as it needs to, and not earlier? Before calling persistence, at the same time as setting rangeid: https://github.com/dnr/temporal/commit/14ee38d041c3b2d1606b3c79aa40a78d80f7d9d6
It's not a big deal, just seems slightly nicer.

Reply:
either expose shard API like this
```go
func DeleteWorkflow(namespaceID string, workflowID string, runID string) error
```
and within the above function, shard will add shard ID & range ID, then call persistence layer

or caller should set all variables
otherwise, we will never be sure which variable is set or not.
similar bug was fixed before: #2056 

I do agree range ID & task ID are not handled like above, we probably should improve <-

Reply:
Use WorkflowKey as input then we don't have this confusion.

Reply:
Well, we avoid the question in this PR, but still an open question for which style we prefer